pkg/autoscaler/autoscaler.go

```diff
@@ -114,92 +110,60 @@ func (agg *totalAggregation) aggregate(stat Stat) {
 	agg.probeCount++
 }
 
-// The number of pods that are observable via stats
-// Subtracts the activator pod if its not the only pod reporting stats
-func (agg *totalAggregation) observedPods(now time.Time) float64 {
-	podCount := float64(0.0)
-	for _, pod := range agg.perPodAggregations {
-		podCount += pod.podWeight(now)
-	}
-
-	activatorsCount := len(agg.activatorsContained)
-	// Discount the activators in the pod count.
-	if activatorsCount > 0 {
-		discountedPodCount := podCount - float64(activatorsCount)
-		// Report a minimum of 1 pod if the activators are sending metrics.
-		if discountedPodCount < 1.0 {
-			return 1.0
-		}
-		return discountedPodCount
-	}
-	return podCount
-}
-
-// The observed concurrency of a revision (sum of all average concurrencies of
-// the observed pods)
-// Ignores activator sent metrics if its not the only pod reporting stats
-func (agg *totalAggregation) observedConcurrency(now time.Time) float64 {
-	accumulatedConcurrency := float64(0)
-	activatorConcurrency := float64(0)
-	for podName, perPod := range agg.perPodAggregations {
-		if isActivator(podName) {
-			activatorConcurrency += perPod.calculateAverage(now)
+// The observed concurrency of a revision and the observed concurrency per pod.
+// Ignores activator sent metrics if its not the only pod reporting stats.
+func (agg *totalAggregation) observedConcurrency() (float64, float64) {
+	accumulatedPodConcurrency := 0.0
+	accumulatedRevConcurrency := 0.0
+	activatorConcurrency := 0.0
+	accumulatedProbeCount := int32(0)
+	for _, perPod := range agg.perPodAggregations {
+		if perPod.isActivator {
+			activatorConcurrency += perPod.averagePodConcurrency()
 		} else {
-			accumulatedConcurrency += perPod.calculateAverage(now)
+			accumulatedPodConcurrency += perPod.accumulatedPodConcurrency
+			accumulatedRevConcurrency += perPod.accumulatedRevConcurrency
+			accumulatedProbeCount += perPod.probeCount
 		}
 	}
-	if accumulatedConcurrency < approximateZero {
-		return activatorConcurrency
+	if accumulatedProbeCount != 0 {
+		accumulatedPodConcurrency = accumulatedPodConcurrency / float64(accumulatedProbeCount)
```

Comment:
Added two new vars.

Reply:
+1 This may be a good time to separate the activator's metrics from the pod metrics.

The meaning of `Stat` has changed significantly by adding `AverageRevConcurrency`.  It is now an aggregated data point rather than the measurement of concurrency (queue) on a single pod.  The Activator is generating a `Stat` as a hack to report queue depth by "pretending" to be a single pod.  That worked well enough when the Autoscaler was using Stats to represent single pods.  But now the two metric streams have almost completely diverged--they only share a struct and are special cased throughout the aggregations.

How about something like this?

```golang
// Used by the ServiceScraper.
type Sample struct {
	Time *time.Time
	RequestCount int32 // or AverageRequestsPerSecond float64
	AverageRevConcurrency float64
}
// Used by the Activator
type Stat {
	PodName string
	PendingRequests int32
}

p := endpointsInformer.CountReadyPods()
q := totalRevisionQueueLength.latencyQueueLength() // average of latest Stat for all Activator pods
c := totalAggregation.observedConcurrency()
if q != 0 {
	// Scale to the number of pending requests
	// or the number of current pods (like panic mode).
	return max(p,q)
	// TODO: scale lower when historical request latency is very small
}
// Scale to achieve desired concurrency.
return p * c / target
```
This will dramatically simplify the special case handling inside `totalAggregation`.

Reply:
How about separating them only in the way how them are stored in autoscaler, not the way they are received in autoscaler? If we separate them entirely, we need a new interface to record sample data from scraper, which could make data pipeline complicated.